### DISCUSSION TITLE
textcat model is not updating

### DISCUSSION THREAD

**Discussion by alitsaberi** [explosion/spaCy#9019] [Help: Coding & Implementations] [answered]
Hi everyone,

I'm trying to train a custom text categorizer with spacy v3. But the losses aren't changing through the iterations and in fact they are exactly the same. Here is part of my code:

```
nlp = spacy.load('en_core_web_trf')

config = {
    "threshold": 0.5,
}

textcat = nlp.add_pipe("textcat", config=config, last=True)
for label in binarizer.classes_:
    textcat.add_label(str(label))

optimizer = nlp.initialize()

for i in range(50):
    random.shuffle(TRAIN_DATA)
    losses = {}
    for batch in minibatch(TRAIN_DATA, size=16):
        textcat.update(batch, sgd=optimizer, losses=losses)
    print(i, losses)
```

TRAIN_DATA is a list of examples.

Thanks

**Comment by adrianeboyd** [comment, marked as answer]
There are a number of details to manage in a training loop and the default `trf` pipeline has some config settings that aren't going to work well in this kind of simple training loop. Here are more details in a response to a similar thread, the part about stepping the learning rate schedule is likely relevant:

https://github.com/explosion/spaCy/discussions/7942#discussioncomment-683818

We'd recommend using the train CLI instead, since it manages a lot of the details for you. You can see some examples for `textcat` in the demo projects here:

https://github.com/explosion/projects/tree/v3/pipelines

You can train a textcat component separately (with defaults that work better than the `en_core_web_trf` settings) and add it to `en_core_web_trf` with `source`:

```python
nlp.add_pipe("textcat", source=spacy.load("/path/to/my_custom_textcat_model"))
```

**Reply by alitsaberi** [reply]
Thanks for your reply.

It seems that the learning rate was the problem. I could manage to somehow solve this by setting a learning rate for optimizer:

`optimizer.learn_rate = 0.01`

As mentioned there are a lot details to manage in a training loop and the learning rate is just one of them. Anyway, by setting the learning rate, losses are changing and the textcat model seems to be updating. :)
